### PR TITLE
zebra-state: Add support for temporary sled databases

### LINKED ISSUE #896: Add support for temporary sled databases.
Sled has support for in-memory temporary databases.  We should use this instead of our current tempdir-based solution.

- [ ] Add a new option to `zebra-state::Config`:
```
/// Whether to use an ephemeral, in-memory store instead of writing to disk.
///
/// Set to `false` by default. If this is set to `true`, [`cache_dir`] is ignored and no state
/// is written to disk.  This is useful for testing but is otherwise probably not what you
/// want, unless you intend to load the entire blockchain into memory.
ephemeral: bool,
```

- [ ] Use the `ephemeral` variable to set https://docs.rs/sled/0.34.2/sled/struct.Config.html#method.temporary

- [ ] In `zebra-test`, replace the state `TempDir` configs with a config that sets `ephemeral = true`.
  - We'll need to keep the `TempDir`s for the tests that generate configs

### PR DESCRIPTION
**Pull request by vramana**
Fixes #896

### CODE REVIEW

File: zebra-state/src/lib.rs
Comment (yaahc) [inline zebra-state/src/lib.rs]: beautiful :pray:

**Review by yaahc — CHANGES_REQUESTED** [review]
looks like ephemeral needs to be pub but everything else looks great, thank you!

File: zebra-state/src/lib.rs
Comment (vramana) [reply zebra-state/src/lib.rs]: Fixed

Comment (vramana) [reply zebra-state/src/lib.rs]: Oh the comment was already in the issue mentioned. I didn't write anything new.

Comment (yaahc) [reply zebra-state/src/lib.rs]: ooh, well then, lol.

**Review by yaahc — DISMISSED** [review]
looks good to me, gonna give the other @ZcashFoundation/zebra-team members a chance to review it before merging though. Thank you again.

File: zebra-test/src/command.rs
Comment (oxarbitrage) [inline zebra-test/src/command.rs]: According to the issue description(https://github.com/ZcashFoundation/zebra/issues/896) it seems that if we have `ephemeral = true` in the config we probably dont need `cache_dir` and also we probably dont need to crate the tmp directory.

File: zebra-state/src/lib.rs
Comment (dconnolly) [inline zebra-state/src/lib.rs]: I would personally remove the last line because it can be useful in general for stateless/volatile usage.

```suggestion
    /// is written to disk (meaning the entire blockchain may be loaded into memory).
```

File: zebra-test/src/command.rs
Comment (dconnolly) [reply zebra-test/src/command.rs]: I agree, even if `cache_dir` is correctly ignored, we can elide it here methinks.

**Comment by vramana** [comment]
There seems to be some test failures. I will try to check what they are about later in the evening.

File: zebra-test/src/command.rs
Comment (teor2345) [reply zebra-test/src/command.rs]: As part of the acceptance tests for this feature, we should write a test which makes sure that `cache_dir` is ignored when `ephemeral` is true. This test should set `cache_dir` to a directory inside a new temp dir, then check that zebrad does not use the directory for its state.

Then the other tests can assume that `cache_dir` is ignored. And so they won't need to create the temp dir.

(We will be adding additional state to zebrad when we add wallet support. So a test helps us prevent future bugs.)

File: zebra-state/src/lib.rs
Comment (teor2345) [inline zebra-state/src/lib.rs]: Sled doesn't actually guarantee that ephemeral databases are created in memory:
https://docs.rs/sled/0.34.3/sled/struct.Config.html#method.temporary

In the current version of sled:

If the configured path is changed, sled uses that path, regardless of the temporary setting:
https://docs.rs/sled/0.34.3/src/sled/config.rs.html#250

If the configured path is the default, and the OS is Linux, sled uses an in-memory temporary directory. Otherwise, it uses an on-disk temporary directory:
https://docs.rs/sled/0.34.3/src/sled/config.rs.html#445

```suggestion
    /// Whether to use an ephemeral database.
    ///
    /// Ephemeral databases are stored in memory on Linux, and in a temporary directory on other OSes.
    ///
    /// Set to `false` by default. If this is set to `true`, [`cache_dir`] is ignored.
```

File: zebra-state/src/lib.rs
Comment (teor2345) [inline zebra-state/src/lib.rs]: Question: Should we allow the user to set the temporary directory for ephemeral databases?

What if their default temporary directory doesn't have enough space?
(That's RAM and maybe swap on Linux, or the relevant partition on other OSes.)

If we want them to be able to set the directory, then this code is ok.

Otherwise, we should stop calling `Config::path` when `ephemeral` is true.

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: Since we don't set a cache_dir here, the ephemeral database uses Zebra's default cache dir in all the parallel acceptance tests, causing conflicts.

If we want sled to choose a unique cache_dir, and use a memory cache on Linux, we must stop calling `sled::Config::path` when ephemeral is true. Otherwise, we should supply a unique `cache_dir` in the test config.

File: zebra-test/src/command.rs
Comment (teor2345) [reply zebra-test/src/command.rs]: Let's also add tests that run zebrad with a non-ephemeral state, and check that the state directory is used. (That is, there are some extra files present in the directory after we run zebrad.)

See my review for details.

**Review by teor2345 — CHANGES_REQUESTED** [review]
> There seems to be some test failures. I will try to check what they are about later in the evening.

I had a look at the failures - it looks like the original ticket misunderstood sled's ephemeral feature.

Sled uses the supplied config path, even if `temporary` is set. In Zebra's case, that means that all the acceptance tests use the same path, which causes conflicts, because they run in parallel.

In-memory databases are only used on Linux, and only if the path is not changed. See my review comments for more details.

So looks like we have a few design questions to sort out here. Hopefully we can work them out over the next day or two.

File: zebra-state/src/lib.rs
Comment (vramana) [reply zebra-state/src/lib.rs]: I chose the latter for now. If anyone asks for supporting temporary paths then we can change the behaviour.

**Comment by vramana** [comment]
@teor2345 Thanks for the looking into the test failures :heart:

File: zebrad/tests/acceptance.rs
Comment (yaahc) [inline zebrad/tests/acceptance.rs]: I'd prefer an `Eq` derive over `matches!` for a simple enum like this.

File: zebrad/tests/acceptance.rs
Comment (yaahc) [inline zebrad/tests/acceptance.rs:19]: :+1:

File: zebrad/tests/acceptance.rs
Comment (vramana) [inline zebrad/tests/acceptance.rs:18]: @teor2345 Does this approach to test the persistent database look okay to you? I modified an existing test currently. It passes locally. Should I leave it as is or make a new test?

**Comment by vramana** [comment]
```
failures:

---- batch_flushes_on_max_items stdout ----
The application panicked (crashed).
Message:  assertion failed: timeout(Duration::from_secs(1),
        sign_and_verify(verifier, 100, None)).await.is_ok()
Location: tower-batch/tests/ed25519.rs:133

Backtrace omitted.
Run with RUST_BACKTRACE=1 environment variable to display it.
Run with RUST_BACKTRACE=full to include source snippets.

---- batch_flushes_on_max_latency stdout ----
The application panicked (crashed).
Message:  assertion failed: timeout(Duration::from_secs(1),
        sign_and_verify(verifier, 10, None)).await.is_ok()
Location: tower-batch/tests/ed25519.rs:148

```

Dunno what's causing these test failures. Are they a known issue?

**Comment by yaahc** [comment]
> Dunno what's causing these test failures. Are they a known issue?

Not a known issue as far as I know. It might be that your tweaks to the `zebra-state` config broke them? doing `assert!(result.is_ok());` should definitely be changed though, because that discards the error message which is clearly causing issues. We should change these to use `?` to return the error to end the test if they encounter one.

**Comment by vramana** [comment]
@yaahc  I changed the error reporting.

**Comment by yaahc** [comment]
> @yaahc I changed the error reporting.

awesome, lets see how the CI goes, from your change I noticed that we were only checking the outermost error type, so if either of those inner Results are supposed to be Error this might fail, in which case we will need to change one of those to an `assert!(result.is_err())`, which should be fine because we'd never accidentally discard an error message unlike with `.is_ok()`.

File: zebra-state/src/lib.rs
Comment (teor2345) [reply zebra-state/src/lib.rs]: Yes, I think the simpler implementation is good for now.

(But if we stop ignoring the configured state path in future, that would be a breaking change. So it would be best to add a separate option for the temporary state path.)

Comment (teor2345) [reply zebra-state/src/lib.rs]: > > Otherwise, we should stop calling `Config::path` when `ephemeral` is true.
>
> I chose the latter for now. If anyone asks for supporting temporary paths then we can change the behaviour.

I can't see this change anywhere. Are you planning on making it in a future commit?

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: If we don't set the sled path when ephemeral is true, we won't need to set `cache_dir` here either:
https://github.com/ZcashFoundation/zebra/pull/939/files#r475633200

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs:274]: Looks good!

File: zebrad/tests/acceptance.rs
Comment (teor2345) [reply zebrad/tests/acceptance.rs:18]: I like the enum, it's easier to understand than a bool.

I think it's ok to modify the existing test. We expect `zebrad start` to use a sled database.

Comment (teor2345) [reply zebrad/tests/acceptance.rs:18]: Actually, I think it might be better to have separate `cache_dir` and `ephemeral` tests.

If we have separate tests, it will be easier to diagnose test failures. And it will be easier to maintain our tests - we want to keep testing these features, even if we add or modify tests. (If we test too many things in the one test function, its easy to accidentally delete one of them.)

Here's what the ephemeral test would look like:

> As part of the acceptance tests for this feature, we should write a test which makes sure that cache_dir is ignored when ephemeral is true. This test should set cache_dir to a directory inside a new temp dir, then check that zebrad does not use the directory for its state.

File: zebrad/tests/acceptance.rs
Comment (yaahc) [reply zebrad/tests/acceptance.rs:274]: looks gorgeous

```rust
assert!(cache_dir.read_dir()?.count() > 0);
```

I find this one line to be _extremely_ satisfying. It sings to my error handling brain.

**Review by teor2345 — CHANGES_REQUESTED** [review]
Looks good.

I think we're almost done here, we just need to:
* stop setting `sled::Config::path` when `ephemeral` is true
* add a test that makes sure an ephemeral config doesn't use `cache_dir`

File: zebrad/tests/acceptance.rs
Comment (teor2345) [reply zebrad/tests/acceptance.rs:274]: It's also *very* obvious what it's meant to be testing.

File: zebra-state/src/lib.rs
Comment (vramana) [reply zebra-state/src/lib.rs]: I actually wrote that change but rolled it back due to tes failure. Let me add it back.

File: zebra-state/src/lib.rs
Comment (teor2345) [inline zebra-state/src/lib.rs]: Looks good, but let's avoid duplicating code here.

Otherwise the other options in the ephemeral and path configs could get out of sync, when we add or modify those options.

```suggestion
        let config = sled::Config::default()
            .cache_capacity(self.memory_cache_bytes)
            .mode(sled::Mode::LowSpace);
        if self.ephemeral {
            config.temporary(self.ephemeral)
        } else {
            let path = self.cache_dir.join(net_dir).join("state");
            config.path(path)
        }
```

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: We want to make sure that sled doesn't use the directory - even if the directory is created, that's ok, as long as there is nothing in it.

```suggestion
    // Check that we do not generate cache_dir
    let cache_dir = tempdir.join("state");
    assert!(!cache_dir.exists() || cache_dir.read_dir()?.count() == 0);
```

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs:274]: Can we split this check out into a separate test, which just checks the `cache_dir` option?

See my earlier comment for some background:
https://github.com/ZcashFoundation/zebra/pull/939/files#r475980074

File: zebrad/tests/acceptance.rs
Comment (teor2345) [reply zebrad/tests/acceptance.rs]: Also, can we split this check out into a separate test, which just checks the `ephemeral` option?

See my earlier comment for some background:
https://github.com/ZcashFoundation/zebra/pull/939/files#r475980074

**Review by teor2345 — CHANGES_REQUESTED** [review]
Looks great, just a few tweaks, and we'll be done.

File: zebra-state/src/lib.rs
Comment (teor2345) [reply zebra-state/src/lib.rs]: Looks good now!

**Comment by yaahc** [comment]
@vramana let me know if there's anything I can do to help push this PR over the finish line, this PR is blocking #946 so I'd like to make sure we don't take to long closing this one. If you're not available rn I can take over finishing this PR.

**Comment by vramana** [comment]
@yaahc I will finish this tomorrow. If you need sooner than that, you can take over. No issues at all.

**Comment by yaahc** [comment]
> @yaahc I will finish this tomorrow. If you need sooner than that, you can take over. No issues at all.

tomorrow should be fine, tyty.

File: zebrad/tests/acceptance.rs
Comment (yaahc) [inline zebrad/tests/acceptance.rs]: I think you got the tests mixed up, this one uses `ConfigMode::Ephemeral` but it's called `persistent_mode`

Comment (vramana) [reply zebrad/tests/acceptance.rs]: Oops

Comment (vramana) [reply zebrad/tests/acceptance.rs]: Fixed them

**Review by yaahc — DISMISSED** [review]
Looks great! thank you for another contribution ^_^

File: zebrad/tests/acceptance.rs
Comment (yaahc) [inline zebrad/tests/acceptance.rs:280]: We still want to keep the old version that tests `ConfigMode::Ephemeral` when its configured correctly, can you add back the old version of the test as a separate test and then rename this version to `fn misconfigured_ephemeral_mode()`?

**Review by yaahc — CHANGES_REQUESTED** [review]

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs:211]: Suggestion: Now we have a separate persistent test, we don't need to use persistent for an unrelated test.

```suggestion
    let (tempdir, _guard) = tempdir(ConfigMode::Ephemeral)?;
```

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs:312]: 👍

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: Issue: When ephemeral is set, zebrad should not create the `cache_dir`.

```suggestion
    assert!(!cache_dir.exists());
```

File: zebrad/tests/acceptance.rs
Comment (teor2345) [reply zebrad/tests/acceptance.rs:280]: Thanks, looks good!

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs:315]: We're duplicating the test config defaults here.

That's ok for now, but I've opened #967 to remind us to clean it up in a future PR.

**Review by teor2345 — CHANGES_REQUESTED** [review]
Thanks, this is great!

I just noticed two little tweaks, which I'll apply now, and merge if they pass CI. (They are minor, so I don't want to block the PR on them.)

**Review by teor2345 — DISMISSED** [review]
I'm happy with the PR as it is, but let's wait for @yaahc's final review before merging.

File: zebrad/tests/acceptance.rs
Comment (teor2345) [inline zebrad/tests/acceptance.rs]: @vramana I just read your comment on Discord, and I understand why we were using persistent here. I'll just change it back.

Sorry about that!

```suggestion
    // start caches state, so run one of the start tests with persistent state
    let (tempdir, _guard) = tempdir(ConfigMode::Persistent)?;
```

**Comment by teor2345** [comment]
Thanks so much for this PR, @vramana -  sorry it was a bit more complicated than we thought!

(I just saw @yaahc's comment in Discord asking me to do the final merge, so I'll do it now.)

**Pull request merged**
